stdlib/public/core/EitherSequence.swift

```diff
@@ -0,0 +1,197 @@
+
+  internal var startIndex: Index {
+    switch self {
+    case let .left(s): return .left(s.startIndex)
+    case let .right(s): return .right(s.startIndex)
+    }
+  }
+
+  internal var endIndex: Index {
+    switch self {
+    case let .left(s): return .left(s.endIndex)
+    case let .right(s): return .right(s.endIndex)
+    }
+  }
+
+  internal subscript(position: Index) -> Element {
+    switch (self,position) {
+    case let (.left(s),.left(i)): return s[i]
+    case let (.right(s),.right(i)): return s[i]
+    default: fatalError("_EitherCollecton: Sequence used with other index type")
+    }
+  }
+
+  internal func index(after i: Index) -> Index {
+    switch (self,i) {
+    case let (.left(s),.left(i)): return .left(s.index(after: i))
+    case let (.right(s),.right(i)): return .right(s.index(after: i))
+    default: fatalError("_EitherCollecton: wrong type of index used")
+    }
+  }
+
+  internal func index(
+    _ i: Index,
+    offsetBy distance: Int,
+    limitedBy limit: Index
+  ) -> Index? {
+    switch (self,i,limit) {
+    case let (.left(s),.left(i),.left(limit)):
+      return s.index(i, offsetBy: distance, limitedBy: limit).map { .left($0) }
+    case let (.right(s),.right(i),.right(limit)):
+      return s.index(i, offsetBy: distance, limitedBy: limit).map { .right($0) }
+    default: fatalError("_EitherCollecton: wrong type of index used")
+    }
+  }
+
+  internal func index(_ i: Index, offsetBy distance: Int) -> Index {
+    switch (self,i) {
+    case let (.left(s),.left(i)): return .left(s.index(i, offsetBy: distance))
+    case let (.right(s),.right(i)): return .right(s.index(i, offsetBy: distance))
+    default: fatalError("_EitherCollecton: wrong type of index used")
+    }
+  }
+
+  internal func distance(from start: Index, to end: Index) -> Int {
+    switch (self,start,end) {
+    case let (.left(s),.left(i),.left(j)):
+      return s.distance(from: i, to: j)
+    case let (.right(s),.right(i),.right(j)):
+      return s.distance(from: i, to: j)
+    default: fatalError("_EitherCollecton: wrong type of index used")
+    }
+  }
+}
+
+internal typealias _EitherBidirectionalCollection<
+  L: BidirectionalCollection, R: BidirectionalCollection
+> = _Either<L,R> where L.Element == R.Element
+
+extension _EitherBidirectionalCollection: BidirectionalCollection {
+  internal func index(before i: Index) -> Index {
+    switch (self,i) {
+    case let (.left(s),.left(i)): return .left(s.index(before: i))
+    case let (.right(s),.right(i)): return .right(s.index(before: i))
+    default: fatalError("_EitherCollecton: wrong type of index used")
+    }
+  }
+}
+
+internal typealias _EitherRandomAccessCollection<
+  L: RandomAccessCollection, R: RandomAccessCollection
+> = _Either<L,R> where L.Element == R.Element
+
+extension _EitherRandomAccessCollection: RandomAccessCollection { }
+
+extension _Either {
+  init<T, C: Collection>(
+    _ collection: C
+  ) where Right == AnyCollection<T>, C.Element == T {
```

Comment:
I think we can drop `T` by expressing this as `Right == AnyCollection<C.Element>`